Unmute topic

prakhar1144 opened the pull request:
This carries forward the work in #22923 

Fixes #2517.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -96,39 +97,45 @@ def set_topic_mutes(
+        recipient_id=recipient_id,
+        topic_name=topic_name,
+        defaults={"last_updated": date_added, "visibility_policy": visibility_policy},
     )
```
What's the motivation for the change away from `bulk_create`? I worry about this accidentally reverting the work in 781107308da36f3a5ab70408a0f95acb3b2a82e4.

prakhar1144 replied:
I have reverted it back to `bulk_create`. [Django Docs](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-create) for the new parameters used.

timabbott commented on zerver/tests/test_message_edit.py:
```diff
@@ -1328,7 +1328,7 @@ def notify(user_id: int) -> Dict[str, Any]:
         # This code path adds 9 (1 + 4/user with muted topics) + 1 to
         # the number of database queries for moving a topic.
-        with self.assert_database_query_count(19):
+        with self.assert_database_query_count(21):
```
Why does this add 2 database queries -- is that the `bulk_create` having been more efficient?

If we need these queries, we need to update the comment, but I think it'd be better to avoid adding more queries if at all possible.

prakhar1144 replied:
> is that the bulk_create having been more efficient?

Yes.

After reverting back to `bulk_create`, all the extra database queries added up in the previous commit are now reverted.

timabbott commented on zerver/tests/test_topic_visibility_policies.py:
```diff
@@ -223,3 +228,49 @@ def test_muted_topic_remove_invalid(self) -> None:
+        hamlet_date_unmuted = UserTopic.objects.filter(
+            user_profile=hamlet, visibility_policy=UserTopic.UNMUTED
+        )[0].last_updated
+        self.assertTrue(timezone_now() - hamlet_date_unmuted <= timedelta(seconds=100))
```
It's best to never use real time in tests. The 100s delay you have here should be fine, but a very large portion of flaky tests result from using real time. Just make up a fixed value for `date_added` to use inside `set_topic_visibility_for_user` and check that the value is as expected.

prakhar1144 replied:
**To Do**

prakhar1144 replied:
I have added a [prep commit](https://github.com/zulip/zulip/pull/24236/commits/17be5c18ebcfaf5afcc8acc6ee4cad2a9a4b6927) to replace the existing pattern of using real time. (in the existing tests in `test_muting_topic.py`)

Newer tests follow the updated pattern of using fixed value for `date_added`.

timabbott commented on zerver/actions/user_topics.py:
```diff
@@ -52,10 +56,10 @@ def do_unmute_topic(
     # updated for correctness.
     try:
-        remove_topic_mute(user_profile, stream.id, topic)
+        remove_topic_visibility_policy(user_profile, stream.id, topic)
```
For this one, it's possible that we will eventually want a common code path where you just call `add_topic_visibility_policy`, with a `visibility_policy` of `VISIBILITY_POLICY_INHERIT`; but we can leave that sort of refactor for a future PR.

prakhar1144 replied:
After [CZO discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1501281),

It is implemented in this PR itself, `set_user_topic_visibility_policy` is the common function.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -205,6 +205,7 @@ def get_recipient_info(
         # of this function for different message types.
         assert stream_topic is not None
         user_ids_muting_topic = stream_topic.user_ids_with_visibility_policy(UserTopic.MUTED)
+        user_ids_unmuting_topic = stream_topic.user_ids_with_visibility_policy(UserTopic.UNMUTED)
```
I think we discussed this on the previous PR, but we should do a preparatory refactor to make a `StreamTopicTarget` just query all user IDs with a visibility policy once. Maybe it'd work to just do that query when it is initialized, and have it put the objects returned into a dict of sets that can be accessed with this API, but I'm not particular about the details other than that we avoid duplicate database queries here.

prakhar1144 replied:
I tried this approach.

But the problem is, the dictionary doesn't contains updated data. 
Like, In tests we continuosly update the `visibility_policy`, but dictionary which was created during `StreamTopicTarget` initialization has older data.

But, I think we should probably find a way to optimize it. Becase, in future, with the increase in `visibility_policy` choices. query count will keep increasing.

prakhar1144 replied:
> The StreamTopicTarget structure is relevant for only a single message-send request,

Then we should configure separate smaller tests out of [test_stream_recipient_info](https://github.com/zulip/zulip/pull/24236/commits/cc5a03241b356504cf33e9a3389bab5cad5a1de3#diff-338fc8375bdda13ccdd1d4a31b273f2ee9a121f0fbe7bffe820b8b3d24f43001R1859) (Especially when we are updating visibility_policy) 

It uses the same [StreamTopicTarget](https://github.com/zulip/zulip/blob/main/zerver/tests/test_users.py#L1757) object throughout the tests, resulting in [assertion error](https://github.com/zulip/zulip/pull/24236/commits/cc5a03241b356504cf33e9a3389bab5cad5a1de3#diff-338fc8375bdda13ccdd1d4a31b273f2ee9a121f0fbe7bffe820b8b3d24f43001L1870) (after we change visibility policy)

abhijeetbodas2001 replied:
No, what I meant was, there's no need to store the result of the DB query in `StreamTopicTarget`.
It should be suffient to just modify the DB query so that all the data can be fetched by a single query instead of 2 separate ones (as explained in my comment above, last paragraph)

prakhar1144 replied:
So, we are handling this in the following way:

A helper function [`get_visibility_policy_to_user_ids`](https://github.com/zulip/zulip/pull/24236/commits/33506f654c5c06c9419ae2a5185e059979bd2320#diff-d803e860e5ff2068cb51aab642015247be71eb4045444a4dcb82e86486f22a4dR156) is added which returns a dictionary of the form
```
(visibility_policy=>set(user_ids))
```

The function queries the database only once.

and finally we can just do:
```
     visibility_policy_to_user_ids = get_visibility_policy_to_user_ids(stream_topic)
     user_ids_muting_topic = visibility_policy_to_user_ids[UserTopic.MUTED]
     user_ids_unmuting_topic = visibility_policy_to_user_ids[UserTopic.UNMUTED]
```

prakhar1144 replied:
[Updated Approach](https://github.com/zulip/zulip/pull/24236#discussion_r1105304217) -- after #24196

timabbott commented on zerver/tests/test_event_queue.py:
```diff
@@ -608,6 +608,33 @@ def test_stream_push_notify_stream_specific_setting_with_muted_topic(self) -> No
+        self,
+    ) -> None:
+        # Push notification should be sent
+        self.change_subscription_properties({"push_notifications": True, "is_muted": True})
```
Might make sense to do `email_notifications` true as well?

abhijeetbodas2001 replied:
I think it is not necessary to set either `stream.push_notifs` or `stream.email_notifs`. We already test elsewhere than notifications are not sent when `stream.is_muted = True` (even if `stream.push_notifications = True | stream.email_notifications = True`).
So here, just setting `stream.is_muted = True` and then checking that notifications are still sent if the topic is unmuted should be sufficient, I think.

prakhar1144 replied:
@abhijeetbodas2001 right!

I have updated the test.

prakhar1144 replied:
**4 new test cases are added.**

[Details (comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1097780584)

timabbott commented:
We also need an API feature level bump and `Changes` entry in the API changelog; see the new feature tutorial and https://zulip.readthedocs.io/en/latest/documentation/api.html for how to do that correctly.

While reviewing this, I noticed that `build_topic_mute_checker`  has a query that pretty clearly should extend its WHERE clause to look at `visibility_policy=MUTED`. I determined that function was dead code and removed it in 22712f634d85102613ae1b0adaa678898d65abc8.

Otherwise, I think this is pretty close to being ready to merge.

timabbott commented:
The very first UI thing we'll need to do after merging the API will be to add an option to the topic actions popover in the left sidebar called "Unmute topic" that is only displayed if  `page_params.development_environment` is True and will set this flag -- if you can push a commit for that in this PR, that'd be helpful, or it can be in immediate follow-up. 

Once we have that implemented, it'll be possible to work on the ~8 different parts of the frontend UI that will want to be different if a topic is unmuted -- the key thing is that first UI step will make it convenient to test those changes individually on top of `main`, so we can do those as several independent PRs that each can merge once they're individually ready.

alya commented:
> The very first UI thing we'll need to do after merging the API will be to add an option to the topic actions popover in the left sidebar called "Unmute topic" that is only displayed if `page_params.development_environment` is True and will set this flag -- if you can push a commit for that in this PR, that'd be helpful, or it can be in immediate follow-up.

I filed this as #24244. Not marked as "help wanted", as it's blocked on this PR, and in any case it would probably be best for @prakhar1144 to pick it up.

abhijeetbodas2001 commented on zerver/tests/test_users.py:
```diff
@@ -1852,6 +1852,32 @@ def test_stream_recipient_info(self) -> None:
+        self.assertEqual(info["stream_push_user_ids"], {hamlet.id})
+
+        # Now unmute the stream and remove topic visibility_policy.
+        sub.is_muted = False
```
Not in this PR, but it would probably be nice to split this already 200+ lines long test into multiple smaller tests (one for each scenario).

abhijeetbodas2001 reviewed (commented):
Thanks for rebasing this! Really happy to see this moving forward :) I left some comments after reading the code.

abhijeetbodas2001 commented:
I mentioned this in the earlier PR as well (https://github.com/zulip/zulip/pull/22923#pullrequestreview-1272342360), but I think it would be better to use the `user_topic` string than `topic_visibility_policy`/`visibility_policy` in line with the Model name and file names. It is also more concise. (For example, `remove_user_topic` instead of `remove_topic_visibility_policy`). Thoughts?

abhijeetbodas2001 commented:
> We also need an API feature level bump and `Changes` entry in the API changelog; see the new feature tutorial and https://zulip.readthedocs.io/en/latest/documentation/api.html for how to do that correctly.

This PR doesn't affect the API yet, right? This is only internal changes. We can bump the API level when we add new URLs and views for this feature, right?

prakhar1144 commented on zerver/actions/user_topics.py:
```diff
@@ -15,6 +15,37 @@
 
 
+def do_set_unmuted_visibility(
+    user_profile: UserProfile,
```
I think, we don't need all three functions here i.e.
* `do_set_unmuted_visibility`
* `do_unmute_topic` 
* `do_mute_topic`

Instead we can convert `do_mute_topic` into `do_set_user_topic` which as suggested by you : 
> will take in visibility_policy as parameter and thus can be reused later when we add support for other visibility policies

And `do_unmute_user` converted into `do_remove_user_topic` which will delete the `user_topic` row.
> when visibility_policy = INHERITED

We can omit `do_set_unmuted_visibility` and future addition of `do_set_follow_visibility` following the above discussed approach.

sounds good?

prakhar1144 replied:
A generic function named `do_set_user_topic_visibility_policy` has been added.

timabbott commented:
> This PR doesn't affect the API yet, right? This is only internal changes.

Ahh, that's right, since it's not actually possible to create a UserTopic with the new visibility policy yet. 

>  but I think it would be better to use the user_topic string

Agreed!

prakhar1144 commented on zerver/tests/test_topic_visibility_policies.py:
```diff
@@ -26,12 +26,13 @@ def test_get_deactivated_muted_topic(self) -> None:
-            date_muted=datetime(2020, 1, 1, tzinfo=timezone.utc),
+            visibility_policy=UserTopic.MUTED,
+            date_added=datetime(2020, 1, 1, tzinfo=timezone.utc),
         )
```
**To Do**

timabbott replied:
I feel like `self.set_user_topic_visibility` doesn't really add much over just calling the actual `do_set_user_topic_visibility_policy` directly -- you still need to pass all 5/6 parameters. I'm flexible here, maybe it's easiest to skip this idea for now and consider it as a potential follow-up as we add more.

prakhar1144 commented on zerver/tests/test_topic_visibility_policies.py:
```diff
@@ -107,29 +108,33 @@ def test_add_muted_topic(self) -> None:
+
+        # with mock.patch("zerver.views.user_topics.topic_is_muted", return_value=False):
+        #     result = self.api_patch(user, url, data)
+        #     self.assert_json_error(result, "Topic already muted")
```
This needs discussion:

In our [`set_user_topic_visibility_policy`](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR108) function,

If the {"user_profile", "stream", "topic_name"} triplet is already present in `UserTopic` table. Update the `visibility_policy`.
Otherwise, create a new entry.

But this leads to the situation where we are not handling `Integrity Error`, i.e. Already Muted.
Of course, `mute_topic` handles this using `topic_is_muted`. But as mentioned here -- race-condition -- how should we handle this?

Fetching + comparing will lead to the addition of extra queries.

abhijeetbodas2001 replied:
Not an answer to the OP, but I did not entirely understand how the earlier works: https://github.com/zulip/zulip/pull/21535#issuecomment-1416697323
Maybe having an answer to that question would give more context as to how the commit in this PR should be.

timabbott replied:
Please start a #backend conversation about this; I replied to https://github.com/zulip/zulip/pull/21535#issuecomment-1416697323 with some additional context. But I think the right answer is that we should ONLY have the database-level integrity check ... but that we might need to add a missing case-insensitive index and then we'd be able to only have the `IntegrityError` check, and get coverage on it through a straightforward "I tried to upload a thing twice" test.

```
zulip=> \d zerver_usertopic;
                                     Table "zulip.zerver_usertopic"
      Column       |           Type           | Collation | Nullable |             Default              
-------------------+--------------------------+-----------+----------+----------------------------------
 id                | integer                  |           | not null | generated by default as identity
 topic_name        | character varying(60)    |           | not null | 
 recipient_id      | integer                  |           | not null | 
 stream_id         | integer                  |           | not null | 
 user_profile_id   | integer                  |           | not null | 
 last_updated      | timestamp with time zone |           | not null | 
 visibility_policy | smallint                 |           | not null | 
Indexes:
    "zerver_mutedtopic_pkey" PRIMARY KEY, btree (id)
    "zerver_mutedtopic_recipient_id_e1901302" btree (recipient_id)
    "zerver_mutedtopic_stream_id_acbff20e" btree (stream_id)
    "zerver_mutedtopic_stream_topic" btree (stream_id, upper(topic_name::text))
    "zerver_mutedtopic_user_profile_id_4f8a692c" btree (user_profile_id)
    "zerver_mutedtopic_user_profile_id_stream_i_2cb30f72_uniq" UNIQUE CONSTRAINT, btree (user_profile_id, stream_id, topic_name)
    "zerver_usertopic_stream_topic_user_visibility_idx" btree (stream_id, topic_name, visibility_policy, user_profile_id)
    "zerver_usertopic_user_visibility_idx" btree (user_profile_id, visibility_policy, stream_id, topic_name)
```

But since that may be a bit of a project, you can probably unblock this by just using `mock.patch` as shown in the commented-out test to make the first check inoperative, and then remove the extra testing code when we fix the index.

@abhijeetbodas2001 want to open an issue to track improve the `UserTopic` database-level constraint?

prakhar1144 replied:
Tim Abbott Said:
> But I think the right answer is that we should ONLY have the database-level integrity check

I think we should only have `topic_is_muted()` test -- test using Django.
We should **remove** database-level integrity check.

**Reason:**
Earlier in [set_user_topic_visibility_policy()](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR108), we were using:

```python
    UserTopic.objects.bulk_create(
        [
            UserTopic(
                user_profile=user_profile,
                stream_id=stream_id,
                recipient_id=recipient_id,
                topic_name=topic_name,
                last_updated=date_muted,
                visibility_policy=UserTopic.MUTED,
            ),
        ],
        ignore_conflicts=ignore_duplicate,
    )
 ```
It creates a new object or raises an Integrity Error.

Now, we are planning to use: (b/c we have now other visibility policies, other than just MUTE)
```python
        UserTopic.objects.bulk_create(
            [
                UserTopic(
                    user_profile=user_profile,
                    stream_id=stream_id,
                    recipient_id=recipient_id,
                    topic_name=topic_name,
                    last_updated=date_added,
                    visibility_policy=visibility_policy,
                ),
            ],
            ignore_conflicts=ignore_duplicate,
            update_conflicts=update_duplicate,
            update_fields=["visibility_policy"],
            unique_fields=["user_profile", "stream", "topic_name"],
        )
 ```
It creates or updates (if **{user_profile, stream, topic_name}** triplet already exists).

[Django Docs](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-create) -- related to new parameters being used.

DB will no longer raise `Integrity Error`

@timabbott your views on this ?

prakhar1144 replied:
[CZO Discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1506164)

timabbott replied:
This was resolved via #24354.

abhijeetbodas2001 reviewed (commented):
Thanks for updating the PR! I left some comments on the last commit.
For the earlier commits, I think, strategically, it would be much better to first update the functions which are *being called* (the inner functions in `/lib`) and after that change the *callers* (functions in `/actions`).
That would lessen the churn and give a smaller diff and a much more readable commit series I think.

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -239,6 +240,8 @@ def should_send(setting: str, row: Dict[str, Any]) -> bool:
             # This implements the structure that the UserProfile stream notification settings
             # are defaults, which can be overridden by the stream-level settings (if those
             # values are not null).
+            if row["user_profile_id"] in user_ids_unmuting_topic:
+                return True
```
```diff
         def should_send(setting: str, row: Dict[str, Any]) -> bool:
             # This implements the structure that the UserProfile stream notification settings
             # are defaults, which can be overridden by the stream-level settings (if those
             # values are not null).
             if row["is_muted"]:
+               if row["user_profile_id"] not in user_ids_unmuting_topic:
+                  return False
             if row["user_profile_id"] in user_ids_muting_topic:
                 return False
             if row[setting] is not None:
                 return row[setting]
             return row["user_profile_" + setting]
   ```
   
   This satisfies the same logic.

abhijeetbodas2001 replied:
Yes, makes sense 👍

prakhar1144 replied:
Updated `should_send` function:
```diff
        def should_send(setting: str, row: Dict[str, Any]) -> bool:
            # This implements the structure that the UserProfile stream notification settings
            # are defaults, which can be overridden by the stream-level settings (if those
            # values are not null).
-            if row["is_muted"]:
+            if row["is_muted"] and row["user_profile_id"] not in user_ids_unmuting_topic:
                return False
            if row["user_profile_id"] in user_ids_muting_topic:
                return False
            if row[setting] is not None:
                return row[setting]
            return row["user_profile_" + setting]
```

But, as mentioned by `abhijeetbodas2001` 
> I think the logic here is going to get very complicated very quickly when we implement the visibility policies. I had tried to extract out this logic into the notification_data.py file so the reading complexity of the message-send codepath is reduced. @timabbott if you could take a look at https://github.com/zulip/zulip/pull/24196, that would be great!

It needs attention from @timabbott

timabbott replied:
OK took a look at #24196; looks good aside from a small tweak to avoid losing the `wildcard_mention_user_ids` optimization.

timabbott replied:
And I think this should be able to be pretty easily rebased on top of that once it's merged; we just might need to pass a map from `user_id => visibility_policy` instead of `user_ids_muting_topics`, or perhaps just have a `user_ids_unmuting_topics` separate data structure; I'm not sure which is cleaner but the latter might be easier to work with.

prakhar1144 replied:
Rebased + added the required logic to [`user_allows_notifications_in_StreamTopic`](https://github.com/zulip/zulip/pull/24236/commits/15e3edc3e71bc3a9ce63373551be04679be44b8b#diff-8319130d25323802a8a47fa49f43d3c5e4c73567aa1e2fc95adb6d4afa68646aR188)

prakhar1144 commented:
@timabbott This PR is ready for having another look from you.

**Note:** 
* Backend-Test is failing due to 2 lines missing coverage.
* I have discussed the reason [here (comment)](https://github.com/zulip/zulip/pull/24236/files#r1096335153) **[help needed]**

**Summary:**
* This carries forward the work in #22923 
* Earlier I rebased #22923 and created this PR.
* Then, we had a discussion around the required changes in the PR
   * [CZO Discussion -- Abhijeet Bodas](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1501281)
   * [CZO Discussion -- Tim Abbott](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1501581)
* I have made the changes, and addressed the initial review comments of `abhijeetbodas2001`

**Following comments need attention**
 * [Should we add 'self.mute_topic' ?](https://github.com/zulip/zulip/pull/24236#discussion_r1096312608)
 * [abhijeetbodas2001 PR to simplify 'should_send' logic](https://github.com/zulip/zulip/pull/24236#discussion_r1096496828)
 * [Integrity error check only by DB](https://github.com/zulip/zulip/pull/21535#issuecomment-1416697323)

prakhar1144 commented on zerver/lib/user_topics.py:
```diff
@@ -96,39 +97,45 @@ def set_topic_mutes(
+        else:
+            row.visibility_policy = visibility_policy
+            row.last_updated = date_added
+            row.save()
```
Since, we reverted back to `bulk_create`, we no longer need this.

prakhar1144 commented on zerver/tests/test_event_queue.py:
```diff
@@ -633,6 +633,37 @@ def test_stream_email_notify_stream_specific_setting_with_muted_stream(self) ->
 
+    def test_stream_notifications_with_unmuted_topic_and_muted_stream(
+        self,
+    ) -> None:
```
**4 new test cases added** for the new feature (as described in the comment above)

timabbott commented:
I merged the first 3 prep commits as the series ending with 906ff9243aa34c5970a8b2c2daf42eb7438dce69.

prakhar1144 commented:
[PR Summary 1.0](https://github.com/zulip/zulip/pull/24236#issuecomment-1419347396)

**PR Summary 2.0**:
* [Updated](https://github.com/zulip/zulip/pull/24236/commits/15e3edc3e71bc3a9ce63373551be04679be44b8b#diff-8319130d25323802a8a47fa49f43d3c5e4c73567aa1e2fc95adb6d4afa68646aR188) the PR to align with the changes in #24196 -- [Discussion GitHub](https://github.com/zulip/zulip/pull/24236#discussion_r1097928190)
* This PR is now **blocked** with the changes in #24354
Once #24354 is merged:
   * I will update [`set_user_topic_visibility`](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR131) to replace `bulk_create` with **a transaction around the whole block to create/update including the check for whether a row exists**. -- [Discussion CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1506165)

<hr>
   
   **Possible Question:**
   **Q.** Why can't we replace `bulk_create` with a transaction around the whole block to create/update including the check for whether a row exists **without merging** #24354 
   **Ans.** Because it will lead to an increase in DB query count, if we replace `bulk_create` without removing `topic_is_mute` check. Also, 24354 adds case insensitive UNIQUE constraint to user topic.

abhijeetbodas2001 commented on zerver/lib/notification_data.py:
```diff
@@ -177,14 +177,18 @@ def get_email_notification_trigger(self, acting_user_id: int, idle: bool) -> Opt
 def user_allows_notifications_in_StreamTopic(
     stream_is_muted: bool,
     topic_is_muted: bool,
+    topic_is_unmuted: bool,
```
I think it would be better to pass the `visibility_policy` of the topic as an argument to this function, since `topic_is_muted` and `topic_is_unmuted` aren't mutually exclusive.

prakhar1144 replied:
@abhijeetbodas2001 

If we use -- `visibility_policy_to_user_ids` dictionary

`user_allows_notifications_in_StreamTopic` in this case:

```python
def user_allows_notifications_in_StreamTopic(
    user_profile_id: int,
    stream_is_muted: bool,
    stream_specific_setting: Optional[bool],
    global_setting: bool,
    visibility_policy_to_user_ids: Dict[int, Set[int]],
) -> bool:
    """
    Captures the hierarchy of notification settings, where visibility policy is considered first,
    followed by stream-specific settings, and the global-setting in the UserProfile is the fallback.
    """
    user_ids_muting_topic = visibility_policy_to_user_ids[UserTopic.MUTED]
    user_ids_unmuting_topic = visibility_policy_to_user_ids[UserTopic.UNMUTED]

    topic_is_muted = user_profile_id in user_ids_muting_topic
    topic_is_unmuted = user_profile_id in user_ids_unmuting_topic

    if stream_is_muted and not topic_is_unmuted:
        return False

    if topic_is_muted:
        return False

    if stream_specific_setting is not None:
        return stream_specific_setting

    return global_setting
```

For passing `visibility_policy`, we would need `user_ids_to_visibility_policy` dict.
In this case:
```python
def user_allows_notifications_in_StreamTopic(
    stream_is_muted: bool,
    visibility_policy: int,
    stream_specific_setting: Optional[bool],
    global_setting: bool,
) -> bool:
    """
    Captures the hierarchy of notification settings, where visibility policy is considered first,
    followed by stream-specific settings, and the global-setting in the UserProfile is the fallback.
    """
    topic_is_muted = visibility_policy == UserTopic.MUTED
    topic_is_unmuted = visibility_policy == UserTopic.UNMUTED

    if stream_is_muted and not topic_is_unmuted:
        return False

    if topic_is_muted:
        return False

    if stream_specific_setting is not None:
        return stream_specific_setting

    return global_setting
```
<hr>

**My question (possible cons):**
Is there any downside associated with the dict `user_ids_to_visibility_policy`? -- Will it create any issue on prod, since user_ids will be pretty large in number?

**Pros**: Second approach reduces the TC associated with:
```
    topic_is_muted = user_profile_id in user_ids_muting_topic
    topic_is_unmuted = user_profile_id in user_ids_unmuting_topic
```
from O(n) to O(1)

abhijeetbodas2001 replied:
Maybe it would be better to create a mapping of `user_id` -> `visibility_policy` instead of the other way round. That would allow us to pass in the visilibilty ppolicy value for the concerned user to `user_allows_notifications_in_StreamTopic`.

My point was that `user_allows_notifications_in_StreamTopic` should not have to worry about what `user_id`s etc are involved. It should only look at the notification settings and give out the result. That would be a much cleaner interface -- what do you think?

abhijeetbodas2001 replied:
> Maybe it would be better to create a mapping of `user_id` -> `visibility_policy` instead of the other way round. That would allow us to pass in the visilibilty ppolicy value for the concerned user to `user_allows_notifications_in_StreamTopic`.

In that case, we could avoid creating the `user_ids_muting_topic` and `user_ids_unmuting_topic` structures altogether, right?

prakhar1144 replied:
> That would be a much cleaner interface -- what do you think?

**Agreed!**

My previous comment:
> My question (possible cons):
Is there any downside associated with the dict user_ids_to_visibility_policy? -- Will it create any issue on prod, since user_ids will be pretty large in number?

> Pros: The second approach reduces the TC associated with:
  
      topic_is_muted = user_profile_id in user_ids_muting_topic
      topic_is_unmuted = user_profile_id in user_ids_unmuting_topic
> from O(n) to O(1)

I researched a bit regarding the dictionary's performance with its size. -- Since the max subscribers to any stream in CZO is  ~ 24k (other realms would have near-about the same numbers). So, practically our query set size wrt {stream, topic} pair would be way too small. So size won't create any problem at this scale.

**I will refactor this with the `user_ids_to_visibility_policy` map approach.**

prakhar1144 replied:
@abhijeetbodas2001
**PR updated.**

prakhar1144 replied:
@abhijeetbodas2001 
Updated the PR

abhijeetbodas2001 commented on zerver/actions/message_send.py:
```diff
@@ -158,6 +158,20 @@ def render_incoming_message(
 
 
+def get_visibility_policy_to_user_ids(stream_topic: StreamTopicTarget) -> Dict[int, Set[int]]:
+    stream_id: int = stream_topic.stream_id
```
Why not have this as a method in the `StreamTopicTarget` class itself?

abhijeetbodas2001 replied:
Also see my other comment at https://github.com/zulip/zulip/pull/24236#discussion_r1105304217. It might be better to have this function return the mapping in the other way (`user_id` -> `visibility_policy`, that is)

prakhar1144 replied:
> Why not have this as a method in the `StreamTopicTarget` class itself?

Done

abhijeetbodas2001 commented on zerver/lib/notification_data.py:
```diff
@@ -176,15 +176,21 @@ def get_email_notification_trigger(self, acting_user_id: int, idle: bool) -> Opt
+
+    if stream_is_muted and not topic_is_unmuted:
+        return False
+
```
IMO we should probably have the `visibility_policy == UserTopic.MUTED` as part of the `if` statement itself, instead of creating separate `bool`.

Since `topic_is_muted` and `topic_is_unmuted` are not mutually exclusive, to a new reader it might be confusing (as in, shouldn't it be that `topic_is_muted == !topic_is_unmuted`?). Explicitly using the values from the model would prompt the reader to go to the model to read the definitions of `UNMUTED` etc, where we have comments explaining what these mean.

abhijeetbodas2001 replied:
More broadly, I think it would be nice to avoid using the terms "muted"/"unmuted" anywhere in the code. `visibility_policy = UserTopic.UNMUTED` is much more explicit.
(For this reason, it is better to have a single function (which we have now) instead of having functions like `do_mute_topic`/`do_unmute_topic` etc)

prakhar1144 replied:
Makes sense -- Updated the PR.

abhijeetbodas2001 reviewed (commented):
I looked at the last commit -- this looks great! I'm kinda nitpicking in the comment below, so feel free to ignore it if it doesn't make sense :)

prakhar1144 commented:
@timabbott 
This PR is ready for having another look from you.

[**PR Summary**](https://github.com/zulip/zulip/pull/24236#issuecomment-1426987581)

ps: It is blocked with the changes in #24354

timabbott commented:
#24354 is now merged!

prakhar1144 commented:
Rebased and updated the PR.

### Help Required
* CI gives `no-coverage` error for [this line](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR142). -- I have added the [following code](https://github.com/zulip/zulip/pull/24236/files#diff-dca29140ac4182499e88c086202b897cca41c670b6163288b76f582569528402R131) for test coverage, but the issue persists.

### FYI
* There is an **increase in db query count** in the [message-edit code path](https://github.com/zulip/zulip/pull/24236/files#diff-3a0986e298aade053cf2a8360758507bc4fed8c3d1fe341724e6c7fd0dd720a6).
> Reason:
* Earlier, ['add_topic_mute' ](https://github.com/zulip/zulip/blob/main/zerver/lib/user_topics.py#L107)(now, [set_user_topic_visibility_policy](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR110)) either created or raised IntegrityError -- 1 query.
* Now, get_or_create uses **2 queries** in the case of creating new row. 
* We can't use the previous approach, because now we have to handle the case of updating the visibility_policy too.
> Why only the message-edit code path is affected?
* Because we have removed 'topic_is_muted' check in #24354 , so removing one query there and increasing one in 'set_user_topic_visibility_policy' compensates each other. -- message-edit code path didn't involve 'topic_is_muted' check, so no compensation.

<hr>

@abhijeetbodas2001 can you please help me with the `no coverage` error, mentioned above?

abhijeetbodas2001 commented on zerver/lib/user_topics.py:
```diff
@@ -95,39 +96,46 @@ def set_topic_mutes(
+    if not created:
+        if row.visibility_policy == visibility_policy and not ignore_duplicate:
+            raise IntegrityError  # nocoverage
+        else:
```
Hmm, this pattern seems off to me. Why should we need to raise an `IntegrityError` ourselves? Isn't that the database's work?

abhijeetbodas2001 replied:
Why not `raise JsonableError(_("Topic is not muted"))` here, instead of `raise IntegrityError`? That seems more natural to me, although it may not solve the coverage issue.

abhijeetbodas2001 replied:
Maybe its best to not raise anything (to avoid over-complicating things inside a transaction)?

```python
        if not created:
            if row.visibility_policy == visibility_policy:
                if ignore_duplicate:
                    return

            row.visibility_policy = visibility_policy
            row.last_updated = date_added
            row.save(update_fields=["visibility_policy", "last_updated"])
```
Could you check if that works?

prakhar1144 replied:
Ah, actually the `no covereage` error was due to not using **case insensitive** lookup for topic_name in 'get_or_create'
Updating it, fixes the issue.

```diff
get_or_create(
.
-`topic_name=topic_name`
+`topic_name__iexact=topic_name
.
```

Now, addressing the other concerns:
1. If we use the following code, [this piece of test](https://github.com/zulip/zulip/pull/24236/files#diff-dca29140ac4182499e88c086202b897cca41c670b6163288b76f582569528402R126) will fail.
```
        if not created:
            if row.visibility_policy == visibility_policy:
                if ignore_duplicate:
                    return

            row.visibility_policy = visibility_policy
            row.last_updated = date_added
            row.save(update_fields=["visibility_policy", "last_updated"])
```

abhijeetbodas2001 replied:
The link to the test seems to be broken. Could you paste the test code (which fails) here?

prakhar1144 replied:
In `zerver/tests/test_user_topics.py` [Line-123](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L123)
```python
        assert stream.recipient is not None
        result = self.api_patch(user, url, data)
        
        # Now check that error is raised when attempted to mute an already
        # muted topic. This should be case-insensitive.
        data["topic"] = "VERONA3"
        result = self.api_patch(user, url, data)
        self.assert_json_error(result, "Topic already muted") # this assertion fails

```
DB should raise IntegrityError -- as we removed one of the duplicate check in the previous PR #24354

abhijeetbodas2001 replied:
Ah, that makes sense :)

abhijeetbodas2001 replied:
My concern was that it doesn't seem ergonomic to raise an `IntegrityError` like this artificially. Why not return a usual API error instead?
```diff
diff --git a/zerver/lib/user_topics.py b/zerver/lib/user_topics.py
index fb03bf42fe..febe2cd2c6 100644
--- a/zerver/lib/user_topics.py
+++ b/zerver/lib/user_topics.py
@@ -1,12 +1,14 @@
 import datetime
 from typing import Callable, List, Optional, Tuple, TypedDict

-from django.db import IntegrityError, transaction
+from django.db import transaction
 from django.db.models.query import QuerySet
 from django.utils.timezone import now as timezone_now
+from django.utils.translation import gettext as _
 from sqlalchemy.sql import ClauseElement, and_, column, not_, or_
 from sqlalchemy.types import Integer

+from zerver.lib.exceptions import JsonableError
 from zerver.lib.timestamp import datetime_to_timestamp
 from zerver.lib.topic import topic_match_sa
 from zerver.lib.types import UserTopicDict
@@ -142,7 +144,10 @@ def set_user_topic_visibility_policy(

         if not created:
             if row.visibility_policy == visibility_policy and not ignore_duplicate:
-                raise IntegrityError
+                # Duplicate request case (topic-visibility_policy pair already exists)
+                raise JsonableError(_("Topic already muted"))
+
+            # The request is to just *change* the visibility policy of a topic
...skipping...
 from zerver.actions.user_topics import do_set_user_topic_visibility_policy
-from zerver.lib.exceptions import JsonableError
 from zerver.lib.request import REQ, has_request_variables
 from zerver.lib.response import json_success
 from zerver.lib.streams import (
@@ -34,16 +32,13 @@ def mute_topic(
         assert stream_id is not None
         (stream, sub) = access_stream_by_id(user_profile, stream_id)

-    try:
-        do_set_user_topic_visibility_policy(
-            user_profile,
-            stream,
-            topic_name,
-            visibility_policy=UserTopic.MUTED,
-            last_updated=date_muted,
-        )
-    except IntegrityError:
-        raise JsonableError(_("Topic already muted"))
+    do_set_user_topic_visibility_policy(
+        user_profile,
+        stream,
+        topic_name,
+        visibility_policy=UserTopic.MUTED,
+        last_updated=date_muted,
+    )


 def unmute_topic(
```

prakhar1144 replied:
Hmm, that's a better approach. -- Updated the PR

prakhar1144 commented:
@timabbott PR is ready to review.

Rebased and updated the PR -- after #24354 got merged.

PR **[Summary 3.0](https://github.com/zulip/zulip/pull/24236#issuecomment-1438633719)**

prakhar1144 commented on zerver/lib/user_topics.py:
```diff
@@ -95,39 +98,51 @@ def set_topic_mutes(
+    if not created:
+        if row.visibility_policy == visibility_policy and not ignore_duplicate:
+            # Duplicate request case (topic-visibility_policy pair already exists)
+            raise JsonableError(_("Topic already muted"))
```
While re-reviewing, I found that we can't hardcode the error message as "Topic already muted".
It can be "Topic already unmuted", "Topic already followed" etc.

Also, we didn't get this error means, we need to update our tests for unmute-topic to include this case.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -781,25 +782,33 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+                    stream_being_edited,
+                    orig_topic_name,
+                    visibility_policy=UserTopic.VISIBILITY_POLICY_INHERIT,
+                )
```
I think it's safe for this project to be a follow-up, since we won't be offering these new values in the UI yet, but the logic in this `for muting_user in get_users_muting_topic` will want to move all `UserTopic` records, not sure those with the `MUTED` state. And the TODO comment above about bulk updates should ideally be fixed while we're at it.

timabbott replied:
(But maybe we should open an issue for addressing this detail -- can you do that?).

prakhar1144 replied:
Opened #24574 

I'm curious about:
> the TODO comment above about bulk updates should ideally be fixed while we're at it.

Maybe, we would need to write separate `bulk_update()` and `bulk_delete()` query, instead of using `do_set_user_topic_visibility_policy`

timabbott replied:
Yeah, I think something like that -- which is why I suggested we do it as a follow-up; seems like it may be complicated enough to get bogged down.

timabbott commented on zerver/actions/user_topics.py:
```diff
@@ -6,73 +6,58 @@
+    if visibility_policy == UserTopic.VISIBILITY_POLICY_INHERIT:
+        # Note: If you add any new code to this block, the
+        # set_user_topic_visibility_policy call in do_update_message will need to be
+        # updated for correctness.
```
I think this comment is stale and can be deleted; bd04733c0f5a2fe8cbebccc515847a2e96da559c added the  `skip_muted_topics_event` parameter that replaced that code duplication.

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -72,15 +73,15 @@ def mute_topic_for_user(user: UserProfile) -> None:
             user_profile=hamlet, visibility_policy=UserTopic.MUTED
         )[0].last_updated
-        self.assertTrue(timezone_now() - hamlet_date_muted <= timedelta(seconds=100))
+        self.assertTrue(hamlet_date_muted, date_muted)
```
You forgot to change these to `assertEqual` :)

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -211,3 +215,50 @@ def test_muted_topic_remove_invalid(self) -> None:
+        cordelia_date_unmuted = UserTopic.objects.filter(
+            user_profile=cordelia, visibility_policy=UserTopic.UNMUTED
+        )[0].last_updated
+        self.assertTrue(cordelia_date_unmuted, date_unmuted)
```
This test repeats the bugger pattern.

prakhar1144 replied:
`bugger pattern` ? -- Sorry, I'm unaware of this. Could you please add a little detail or references?

ps: I googled + searched zulip docs, but didn't find anything.

timabbott replied:
Sorry, typo, I meant the buggy `assertTrue` pattern.

prakhar1144 replied:
This is resolved

timabbott commented on zerver/tests/test_message_flags.py:
```diff
@@ -23,14 +23,15 @@
 from zerver.lib.test_helpers import get_subscription, timeout_mock
 from zerver.lib.timeout import TimeoutExpiredError
-from zerver.lib.user_topics import add_topic_mute
+from zerver.lib.user_topics import set_user_topic_visibility_policy
```
I think it'd be better to have these code paths call `do_set_user_topic_visibility_policy`; the `lib` functions are meant to be internal interfaces for just changing the database (but not sending events), and shouldn't generally be called elsewhere.

I'll propose changing their name as well in another comment.

prakhar1144 replied:
The current state of codebase doesn't follow this pattern.

Added a prep commit (2nd commit) -- This commit makes sure, `add_topic_mute` and `remove_topic_mute` is not directly called, instead `do_mute_topic` and `do_unmute_topic` is used.

Later, commits replaces `do_mute_topic` and `do_unmute_topic` with `do_set_user_topic_visibility_policy`.
Hence,
> it'd be better to have these code paths call do_set_user_topic_visibility_policy

is implemented.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -95,39 +98,58 @@ def set_topic_mutes(
 
-def add_topic_mute(
+@transaction.atomic
+def set_user_topic_visibility_policy(
```
Having this function and another with the same name with `do_` at the front is pretty confusing. I think it's possible the best answer would be to inline this into its caller. But an alternative would be to call this something like `set_user_topic_visibility_policy_in_database`

prakhar1144 replied:
I think renaming it to `set_user_topic_visibility_policy_in_database` is the better option -- in favour of readability.

Let me know, if keeping it inline makes more sense. -- (Minor refactoring would be required)

timabbott replied:
Yeah that rename sounds good.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -95,39 +98,58 @@ def set_topic_mutes(
 
 
-def add_topic_mute(
+@transaction.atomic
```
Why don't we need `savepoint=False` on this `transaction.atomic`? I thought this function was called from the message edit code path -- is that outer code path not using a transaction?

prakhar1144 replied:
I read about savepoint and durable params. -- Yes, we need savepoint=False here.

While implementing it, [test_muted_topic_remove_invalid](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L187)  is failing.
I spent 3-4 hours debugging this, no success.

This is the traceback:
```shell
======================================================================
ERROR: test_muted_topic_remove_invalid (zerver.tests.test_user_topics.MutedTopicsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/tests/test_user_topics.py", line 192, in test_muted_topic_remove_invalid
    result = self.api_patch(user, url, data)
  File "/srv/zulip/zerver/lib/test_classes.py", line 934, in api_patch
    return self.client_patch(
  File "/srv/zulip/zerver/lib/test_helpers.py", line 404, in wrapper
    result = f(self, url, info, **kwargs)
  File "/srv/zulip/zerver/lib/test_classes.py", line 314, in client_patch
    result = django_client.patch(url, encoded, follow=follow, secure=secure, **extra)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 920, in patch
    response = super().patch(
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 495, in patch
    return self.generic("PATCH", path, data, content_type, secure=secure, **extra)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 541, in generic
    return self.request(**r)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 810, in request
    self.check_exception(response)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 663, in check_exception
    raise exc_value
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/zulip/zerver/lib/rest.py", line 39, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/srv/zulip/zerver/lib/rest.py", line 203, in rest_dispatch
    return target_function(request, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 816, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 762, in _wrapped_func_arguments
    raise err
  File "/srv/zulip/zerver/decorator.py", line 759, in _wrapped_func_arguments
    return view_func(request, user_profile, *args, **kwargs)
  File "/srv/zulip/zerver/lib/request.py", line 458, in _wrapped_req_func
    return req_func(request, *args, **kwargs)
  File "/srv/zulip/zerver/views/user_topics.py", line 83, in update_muted_topic
    unmute_topic(
  File "/srv/zulip/zerver/views/user_topics.py", line 56, in unmute_topic
    stream = access_stream_for_unmute_topic_by_id(user_profile, stream_id, error)
  File "/srv/zulip/zerver/lib/streams.py", line 523, in access_stream_for_unmute_topic_by_id
    stream = Stream.objects.get(id=stream_id, realm_id=user_profile.realm_id)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 646, in get
    num = len(clone)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 376, in __len__
    self._fetch_all()
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 1867, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 87, in __iter__
    results = compiler.execute_sql(
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1398, in execute_sql
    cursor.execute(sql, params)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 83, in _execute
    self.db.validate_no_broken_transaction()
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 520, in validate_no_broken_transaction
    raise TransactionManagementError(
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

prakhar1144 replied:
**Update: The issue is resolved.**

**Still it needs discussion:**

The error was due to the order of execution of code in the test.

When this piece of code executes earlier in [test_muted_topic_remove_invalid](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L187)
```
        data = {"stream": stream.name, "topic": "BOGUS", "op": "remove"}
        result = self.api_patch(user, url, data)
        self.assert_json_error(result, "Topic is not muted")
```
It results in the above mentioned error (traceback posted in the previous comment).

**Putting it at the bottom of the test method, resolves the issue.**

**Reason:**
The mentioned piece of code  in the test calls `set_user_topic_visibility_policy_in_database`, which on failure rollbacks the test transaction (because savepoint=False).

Now, when the next piece of code in the test method executes, it leads to the error:
> django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.}}

There are two approaches to resolve this:
1. Placing it to the bottom resolves the issue, because there won't be any further query execution. (Now, rollback doesn't affect anything.)
2. We can update the code to:
```python
# Here savepoint is True
with transaction.atomic():
    data = {"stream": stream.name, "topic": "BOGUS", "op": "remove"}
    result = self.api_patch(user, url, data)
    self.assert_json_error(result, "Topic is not muted")
```

I have used the First approach because of its simplicity. Second approach adds another nested transaction within test.

abhijeetbodas2001 replied:
I think the second approach you describe might be better, since someone adding tests in this file in the future would not be affected by the presence of this test (and they would also know what exactly to do if they are in a similar situation).
In any case, whichever approach we go with, I think we should certainly add a comment explaining why this test needs special attention.

timabbott replied:
It looks like you went with the second approach, with a comment after all --reading the test this lgtm.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -95,39 +98,58 @@ def set_topic_mutes(
     )
 
+    if not created:
+        if row.visibility_policy == visibility_policy and not ignore_duplicate:
```
This would be better written as `if created: return` -- that early return pattern is much more readable.

I'm also not sure whether we should just return in the `row.visibility_policy == visibility_policy and ignore_duplicate` duplicate case -- there's no change, so I think changing `last_updated` is arguably wrong? Doing so as an early return might also make this more readable.

prakhar1144 replied:
I agree with the first point.
> if created: return -- that early return pattern is much more readable.

Regarding the second point
>  I think changing last_updated is arguably wrong? 

We are updating `last_updated` (and visibility_policy) only  when **the request is to update `visibility_policy`**

So we can simplify this to:
```python

if created: return

# Already 'muted', 'unmuted' or 'followed' case. -- duplicate request
if row.visibility_policy == visibility_policy and not ignore_duplicate:
   return JsonableError()

# existing_visibility_policy != new_visibility_policy case -- request to update visibility_policy
row.last_updated = last_updated
row.visibility_policy = visibility_policy
row.save(update_fields=["visibility_policy", "last_updated"])
```

Looks more readable?

prakhar1144 replied:
```python

        if created:
            return

        duplicate_request: bool = row.visibility_policy == visibility_policy

        if duplicate_request and ignore_duplicate:
            return

        if duplicate_request and not ignore_duplicate:
            visibility_policy_string: Dict[int, str] = {
                1: "muted",
                2: "unmuted",
                3: "followed",
            }
            raise JsonableError(
                _("Topic already {}").format(visibility_policy_string[visibility_policy])
            )

        row.visibility_policy = visibility_policy
        row.last_updated = last_updated
        row.save(update_fields=["visibility_policy", "last_updated"])
 ```
 Looks more readable?

timabbott replied:
Yeah, that looks great.

timabbott commented:
OK I think the above is my last batch of review feedback for this PR; let me know when it's all resolved and I can integrate this!

prakhar1144 commented:
This week, I have my college exams, which end on March 3.

So, I will carry forward the work from March 4.

timabbott commented:
Sounds good -- best of luck with the exams!

prakhar1144 commented on zerver/actions/user_topics.py:
```diff
@@ -30,8 +29,12 @@ def do_set_user_topic_visibility_policy(
         except UserTopic.DoesNotExist:
+            # TODO: We need to update this error message to handle cases like:
+            # 'Topic is not muted', 'Topic is not unmuted'.
             raise JsonableError(_("Topic is not muted"))
```
"Topic is not in the requested state"

Does it sound good?

prakhar1144 replied:
OK -- made the changes. 
I will open a thread in CZO.

prakhar1144 commented:
[**Summary 4.0**](https://github.com/zulip/zulip/pull/24236#issuecomment-1439097482)

**Summary 5.0:**

Updates in the recent Push:
* Fix -- `assertTrue` to `assertEqual` [(comment 1)](https://github.com/zulip/zulip/pull/24236#discussion_r1117400093) [(comment 2)](https://github.com/zulip/zulip/pull/24236#discussion_r1117542662)
* prep commit so that lib functions are not called directly in tests [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1125478748) -- second commit
* Rename `set_user_topic_visibility_policy` to `set_user_topic_visibility_policy_in_database`. [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1117547299)
* Add savepoint=False [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1125480682)
* Increase `set_user_topic_visibility_policy_in_database` readability -- [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1117762326)

Comments which need attention:
* bulk_update -- follow-up PR [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1125358080) -- doesn't block this PR
* inline `set_user_topic_in_visibility_policy` [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1125478619) -- doesn't block this PR
* Best way to update test_muted_topic_remove_invalid -- [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1126470636) -- doesn't block this PR
* Update error message -- [(comment)](https://github.com/zulip/zulip/pull/24236#discussion_r1125507810)

@timabbott PR is ready to review.

timabbott commented:
I pushed back to this PR with the following changes -- some very light edits to use the early return pattern and avoid an extra `row` variable, with a comment. I think this is a good checkpoint that we can merge, so marking to merge once CI passes, thanks for all the work on this @akashaviator and @prakhar1144!

``` diff
diff --git a/zerver/lib/user_topics.py b/zerver/lib/user_topics.py
index 6990c24a16..44c77b96cf 100644
--- a/zerver/lib/user_topics.py
+++ b/zerver/lib/user_topics.py
@@ -120,47 +120,49 @@ def set_user_topic_visibility_policy_in_database(
     ignore_duplicate: bool = False,
 ) -> None:
     if visibility_policy == UserTopic.VISIBILITY_POLICY_INHERIT:
-        row = UserTopic.objects.get(
+        # Will throw UserTopic.DoesNotExist if the user doesn't
+        # already have a visibility policy for this topic.
+        UserTopic.objects.get(
             user_profile=user_profile,
             stream_id=stream_id,
             topic_name__iexact=topic_name,
-        )
-        row.delete()
-    else:
-        assert last_updated is not None
-        (row, created) = UserTopic.objects.get_or_create(
-            user_profile=user_profile,
-            stream_id=stream_id,
-            topic_name__iexact=topic_name,
-            recipient_id=recipient_id,
-            defaults={
-                "topic_name": topic_name,
-                "last_updated": last_updated,
-                "visibility_policy": visibility_policy,
-            },
-        )
+        ).delete()
+        return
+
+    assert last_updated is not None
+    (row, created) = UserTopic.objects.get_or_create(
+        user_profile=user_profile,
+        stream_id=stream_id,
+        topic_name__iexact=topic_name,
+        recipient_id=recipient_id,
+        defaults={
+            "topic_name": topic_name,
+            "last_updated": last_updated,
+            "visibility_policy": visibility_policy,
+        },
+    )
+
+    if created:
+        return
 
-        if created:
-            return
-
-        duplicate_request: bool = row.visibility_policy == visibility_policy
-
-        if duplicate_request and ignore_duplicate:
-            return
-
-        if duplicate_request and not ignore_duplicate:
-            visibility_policy_string: Dict[int, str] = {
-                1: "muted",
-                2: "unmuted",
-                3: "followed",
-            }
-            raise JsonableError(
-                _("Topic already {}").format(visibility_policy_string[visibility_policy])
-            )
-        # The request is to just 'update' the visibility policy of a topic
-        row.visibility_policy = visibility_policy
-        row.last_updated = last_updated
-        row.save(update_fields=["visibility_policy", "last_updated"])
+    duplicate_request: bool = row.visibility_policy == visibility_policy
+
+    if duplicate_request and ignore_duplicate:
+        return
+
+    if duplicate_request and not ignore_duplicate:
+        visibility_policy_string: Dict[int, str] = {
+            1: "muted",
+            2: "unmuted",
+            3: "followed",
+        }
+        raise JsonableError(
+            _("Topic already {}").format(visibility_policy_string[visibility_policy])
+        )
+    # The request is to just 'update' the visibility policy of a topic
+    row.visibility_policy = visibility_policy
+    row.last_updated = last_updated
+    row.save(update_fields=["visibility_policy", "last_updated"])
 
 
 def topic_is_muted(user_profile: UserProfile, stream_id: int, topic_name: str) -> bool:
```

timabbott commented:
I think reading the code from this state, along with other more tactical follow-ups, we probably do want to either inline `set_user_topic_visibility_policy_in_database` or reorganize this block in `do_set_user_topic_visibility_policy` to remove the if/else clause and just have a single call to `set_user_topic_visibility_policy_in_database`; it'd be very nice to contain the branching behavior to one place. I think my current temptation is to not inline.

```
if visibility_policy == UserTopic.VISIBILITY_POLICY_INHERIT:                   
    try:                                                                       
        set_user_topic_visibility_policy_in_database(                          
            user_profile, stream.id, topic, visibility_policy=visibility_policy
        )                                                                      
    except UserTopic.DoesNotExist:                                             
        raise JsonableError(_("Nothing to be done"))                           
else:                                                                          
    assert stream.recipient_id is not None                                     
    set_user_topic_visibility_policy_in_database(                              
        user_profile,                                                          
        stream.id,                                                             
        topic,                                                                 
        visibility_policy=visibility_policy,                                   
        recipient_id=stream.recipient_id,                                      
        last_updated=last_updated,                                             
        ignore_duplicate=ignore_duplicate,                                     
    )                                                                          
```

The other big semantic thing that probably deserves discussion in "api design" is what we want the long-term semantics to be around "What if the topic already has the target visibility policy". Often it's a cleaner API to just do nothing than return an error, since, for example, a repeated request due to some client-side retry race would just have both calls 200, rather than the first 200 and then the second fail (and thus generate a user-facing error when in fact the right thing happened). But that would be an API change that we'd want to document, so best to discuss with that group and have it be its own tactical pull request.

Let me know if you need any help figuring out next steps; usually with this sort of thing, it makes sense to try to work through all the easy follow-ups so one doesn't have to keep track of them in parallel with exploring the plan for the next major step in completing the feature itself.

prakhar1144 commented:
Next Steps: [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/Unmute.20Topics/near/1519871)

FYI: @timabbott @abhijeetbodas2001
